client/src/main/java/com/microsoft/durabletask/TaskOrchestrationExecutor.java

```diff
@@ -173,23 +171,34 @@ public <V> Task<List<V>> allOf(List<Task<V>> tasks) {
                     .map(t -> t.future)
                     .toArray((IntFunction<CompletableFuture<V>[]>) CompletableFuture[]::new);
 
-            return new CompletableTask<>(CompletableFuture.allOf(futures).thenApply(x -> {
-                ArrayList<V> results = new ArrayList<>(futures.length);
-
-                // All futures are expected to be completed at this point
-                for (CompletableFuture<V> cf : futures) {
-                    try {
-                        results.add(cf.get());
-                    } catch (Exception ex) {
-                        // TODO: Better exception message than this
-                        // TODO: This needs to be a TaskFailedException or some other documented exception type.
-                        //       https://github.com/microsoft/durabletask-java/issues/54
-                        throw new RuntimeException("One or more tasks failed.", ex);
-                    }
-                }
-
-                return results;
-            }));
+            return new CompletableTask<>(CompletableFuture.allOf(futures)
+                    .thenApply(x -> {
+                        List<V> results = new ArrayList<>(futures.length);
+
+                        // All futures are expected to be completed at this point
+                        for (CompletableFuture<V> cf : futures) {
+                            try {
+                                results.add(cf.get());
+                            } catch (Exception ex) {
+                                results.add(null);
+                            }
+                        }
+                        return results;
+                    })
+                    .exceptionally(throwable -> {
+                        ArrayList<Exception> exceptions = new ArrayList<>(futures.length);
+                        for (CompletableFuture<V> cf : futures) {
+                            try {
+                                cf.get();
+                            } catch (ExecutionException ex) {
+                                exceptions.add((Exception) ex.getCause());
+                            } catch (Exception ex){
+                                exceptions.add(ex);
+                            }
+                        }
+                        throw new CompositeTaskFailedException(exceptions.get(0).getMessage(), exceptions);
```

Comment:
I wonder if it would be better to have a more generic message here. Otherwise, developers may not realize that there could be more than one failure. How about something like this?
```suggestion
                        throw new CompositeTaskFailedException(
                            String.format(
                                "%d out of %d tasks failed with an exception. See the exceptions list for details.", 
                                exceptions.size(),
                                futures.length),
                            exceptions);
```

Reply:
My thought was to inform the developers with the first exception message. But this message is more informative.